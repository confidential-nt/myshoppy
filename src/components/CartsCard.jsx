import React, { useState } from "react";
import { useNavigate } from "react-router-dom";
import { AiOutlineMinusSquare, AiOutlinePlusSquare } from "react-icons/ai";
import { BsTrashFill } from "react-icons/bs";

export default function CartsCard({
  id,
  productInCarts,
  onAddCount,
  onDeleteCount,
  onDeleteCarts,
}) {
  const navigate = useNavigate();
  const [count, setCount] = useState(productInCarts.count);

  const handleAddCount = () => {
    onAddCount.mutate(id);
    setCount((prev) => prev + 1);
  };
  const handleDeleteCount = () => {
    if (count === 1) {
      handleDeleteCarts();
      return;
    }
    onDeleteCount.mutate(id);
    setCount((prev) => prev - 1);
  };
  const handleDeleteCarts = () => {
    onDeleteCarts.mutate(id);
    setCount(0);
  };

  return (
    <>
      {count > 0 ? (
        <li className="flex items-center mb-2 last:mb-0">
          <div
            className="w-1/3 mr-4 basis-1/6 cursor-pointer"
            onClick={() =>
              navigate(`/products/${id}`, {
                state: {
                  product: productInCarts,
                },
              })
            }
            role="img"
            aria-label="go-to-product-detail"
          >
            <img
              src={productInCarts.imageURL}
              alt={productInCarts.name}
              className="w-full h-full"
            />
          </div>
          <div className="flex justify-between flex-grow basis-2/3">
            <div className="flex flex-col justify-center">
              <h3>{productInCarts.name}</h3>
              <span className="uppercase text-shoppypink font-bold">
                {productInCarts.option}
              </span>
              <span>â‚©{productInCarts.price}</span>
            </div>
            <div className="flex">
              <button
                type="button"
                className="mr-1"
                onClick={handleDeleteCount}
                aria-label="deleteCount"
              >
                <AiOutlineMinusSquare />
              </button>
              <span className="mr-1 flex items-center">{count}</span>
              <button
                type="button"
                className="mr-3"
                onClick={handleAddCount}
                aria-label="addCount"
              >
                <AiOutlinePlusSquare />
              </button>
              <button
                type="button"
                onClick={handleDeleteCarts}
                aria-label="deleteCarts"
              >
                <BsTrashFill />
              </button>
            </div>
          </div>
        </li>
      ) : null}
    </>
  );
}
