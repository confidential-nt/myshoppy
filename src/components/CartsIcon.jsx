import React, { useEffect, useState } from "react";
import { Link } from "react-router-dom";
import { AiOutlineShoppingCart } from "react-icons/ai";
import { useUserContext } from "../context/UserContext";
import { useUserRepositoryContext } from "../context/UserRepositoryContext";

export default function CartsIcon() {
  const [number, setNumber] = useState(0);
  const { uid } = useUserContext();
  const { userRepository } = useUserRepositoryContext();

  useEffect(() => {
    if (uid) {
      userRepository.findById(uid).then((user) => {
        if (user) {
          const number = user.carts ? Object.keys(user.carts).length : 0;
          setNumber(number);
        }
      });
    }

    if (uid) {
      userRepository.onUpdateCarts((carts) => {
        setNumber(carts ? Object.keys(carts).length : 0);
      }, uid);
    }
  }, [uid, userRepository]);

  return (
    <Link to="/carts" className="mr-3 lg:mr-7 text-sm lg:text-xl relative">
      {uid && (
        <div className="absolute -right-3 -top-2 text-xs text-white bg-red-500 rounded-full w-4 text-center">
          {number}
        </div>
      )}
      <AiOutlineShoppingCart />
    </Link>
  );
}
