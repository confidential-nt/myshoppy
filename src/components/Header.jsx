import React from "react";
import { Link } from "react-router-dom";
import { LuShoppingBag } from "react-icons/lu";
import { AiOutlineShoppingCart } from "react-icons/ai";

export default function Header() {
  return (
    <header className="w-full flex justify-between items-center pt-1 pr-8 pb-1 pl-2 border">
      <div>
        <Link to="/" className="flex items-center text-shoppypink text-xl">
          <LuShoppingBag className="mr-2" />
          <h1>Shoppy</h1>
        </Link>
      </div>
      <div className="flex items-center font-semibold">
        <Link to="/products" className="text-xs mr-3">
          Products
        </Link>
        <Link to="/carts" className="mr-7 text-xl">
          <AiOutlineShoppingCart />
        </Link>
        <button
          type="button"
          className="bg-shoppypink text-white  pt-1 pr-3 pb-1 pl-3"
        >
          Login
        </button>
      </div>
    </header>
  );
}
