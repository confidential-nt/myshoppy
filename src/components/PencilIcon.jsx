import React, { useEffect, useState } from "react";
import { BsPencilFill } from "react-icons/bs";
import { Link } from "react-router-dom";
import { useUserContext } from "../context/UserContext";
import { useUserRepositoryContext } from "../context/UserRepositoryContext";

export default function PencilIcon() {
  const { uid } = useUserContext();

  const { userRepository } = useUserRepositoryContext();
  const [isAdmin, setIsAdmin] = useState(false);

  useEffect(() => {
    userRepository.findById(uid).then((user) => {
      if (user && user.isAdmin) {
        setIsAdmin(true);
      } else {
        setIsAdmin(false);
      }
    });
  }, [uid, userRepository]);

  return (
    <>
      {isAdmin && (
        <Link to="/products/add" className="mr-3 lg:mr-7 text-sm lg:text-xl">
          <BsPencilFill />
        </Link>
      )}
    </>
  );
}
