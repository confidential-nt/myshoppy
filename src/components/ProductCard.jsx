import React from "react";

export default function ProductCard({ product }) {
  console.log(product);
  return (
    <li className="basis-1/4 mr-3 last:mr-0 cursor-pointer shadow-lg hover:scale-105 transition">
      <div className="w-full">
        <img
          src={product.imageURL}
          alt={product.name}
          className="w-full h-full"
        />
      </div>
      <div className="p-4">
        <div className="flex justify-between">
          <h4>{product.name}</h4>
          <strong className="font-normal">â‚©{product.price}</strong>
        </div>
        <span className="text-sm text-gray-700">{product.category}</span>
      </div>
    </li>
  );
}
