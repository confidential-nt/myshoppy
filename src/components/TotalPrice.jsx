import React from "react";
import { AiFillPlusCircle } from "react-icons/ai";
import { FaEquals } from "react-icons/fa";

export default function TotalPrice({ productInCarts }) {
  const products = Object.entries(productInCarts).map(([k, v]) => v);
  const totalPrice = products.reduce((acc, cur) => {
    return cur.count * cur.price + acc;
  }, 0);
  const deliveryFee = 3000;

  return (
    <div className="flex flex-col w-10/12 mr-auto ml-auto pt-10">
      <div className="flex justify-around items-center mb-10">
        <div className="flex flex-col items-center">
          <span>상품 총액</span>
          <strong className="text-shoppypink">₩{totalPrice}</strong>
        </div>
        <span>
          <AiFillPlusCircle />
        </span>
        <div className="flex flex-col items-center">
          <span>배송비</span>
          <strong className="text-shoppypink">₩{deliveryFee}</strong>
        </div>
        <span>
          <FaEquals />
        </span>
        <div className="flex flex-col items-center">
          <span>총가격</span>
          <strong className="text-shoppypink">
            ₩{totalPrice + deliveryFee}
          </strong>
        </div>
      </div>
      <button className="bg-shoppypink text-white text-xl pt-1 pb-1">
        주문하기
      </button>
    </div>
  );
}
