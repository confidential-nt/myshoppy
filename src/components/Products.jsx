import React, { useEffect, useState } from "react";
import ProductCard from "./ProductCard";

export default function Products({ productRepository }) {
  const [products, setProducts] = useState();
  useEffect(() => {
    productRepository.findAll(setProducts);
  }, []);
  return (
    <ul className="flex mt-4 pr-2 pl-3">
      {products &&
        Object.entries(products).map(([key, value]) => (
          <ProductCard key={key} product={value} id={key} />
        ))}
    </ul>
  );
}
