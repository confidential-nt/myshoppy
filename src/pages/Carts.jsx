import React, { useEffect } from "react";
import { useQuery, useMutation, useQueryClient } from "@tanstack/react-query";
import { useNavigate } from "react-router-dom";
import { useUserContext } from "../context/UserContext";
import { useUserRepositoryContext } from "../context/UserRepositoryContext";
import CartsCard from "../components/CartsCard";
import TotalPrice from "../components/TotalPrice";

async function productsInCarts(uid, userRepository, productRepository) {
  const user = await userRepository.findById(uid);
  if (!user) return null;
  if (!user.carts) return null;

  const keys = Object.keys(user.carts);

  return productRepository.findAll().then((products) => {
    const productList = Object.entries(products);
    const productInCarts = productList
      .filter(([k, v]) => keys.includes(k))
      .map(([k, v]) => [
        k,
        {
          ...v,
          option: user.carts[k].option,
          count: user.carts[k].count,
        },
      ]);
    return Object.fromEntries(productInCarts);
  });
}

export default function Carts({ productRepository }) {
  const { uid } = useUserContext();
  const { userRepository } = useUserRepositoryContext();

  const {
    isLoading,
    error,
    data: carts,
  } = useQuery(
    ["carts", uid],
    () => productsInCarts(uid, userRepository, productRepository),
    {
      staleTime: 1000 * 60 * 1,
    }
  );

  const queryClient = useQueryClient();
  const navigate = useNavigate();

  const onAddCount = useMutation(
    (productId) => userRepository.updateCount(uid, productId, 1),

    {
      onSuccess: () => queryClient.invalidateQueries(["carts", uid]),
    }
  );

  const onDeleteCount = useMutation(
    (productId) => userRepository.updateCount(uid, productId, -1),

    {
      onSuccess: () => queryClient.invalidateQueries(["carts", uid]),
    }
  );

  const onDeleteCarts = useMutation(
    (productId) => {
      userRepository.deleteCarts(uid, productId);
    },
    {
      onSuccess: () => queryClient.invalidateQueries(["carts", uid]),
    }
  );

  useEffect(() => {
    if (!uid) {
      navigate("/");
    }
  }, [uid, navigate]);

  return (
    <>
      <h1 className="pt-5 pb-3 font-bold text-center">내 장바구니</h1>
      <ul className="pt-4 pb-4 pl-6 pr-6 border-t border-b w-11/12 mr-auto ml-auto">
        {isLoading ? (
          <p>is loading...</p>
        ) : error ? (
          <p>Something is wrong</p>
        ) : !carts ? null : (
          Object.entries(carts).map(([k, v]) => (
            <CartsCard
              key={k}
              id={k}
              productInCarts={v}
              onAddCount={onAddCount}
              onDeleteCount={onDeleteCount}
              onDeleteCarts={onDeleteCarts}
            />
          ))
        )}
      </ul>
      {isLoading ? null : error ? null : !carts ? null : (
        <TotalPrice productInCarts={carts} />
      )}
    </>
  );
}
