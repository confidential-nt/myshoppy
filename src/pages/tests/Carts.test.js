import { render, screen, waitFor } from "@testing-library/react";
import {
  withAllUserContexts,
  withReactQuery,
  withRouter,
} from "../../tests/utils";
import { Route } from "react-router-dom";
import Carts from "../Carts";
import { fakeProducts } from "../../tests/products";
import { fakeAdminUser } from "../../tests/user";
import * as TotalPrice from "../../components/TotalPrice";

jest.mock("../../components/TotalPrice");

describe("Carts", () => {
  const mockTotalPriceProp = jest.fn();

  TotalPrice.default = function MockTotalPrice({ productInCarts }) {
    mockTotalPriceProp(productInCarts);
    const products = Object.entries(productInCarts).map(([k, v]) => v);
    const totalPrice = products.reduce((acc, cur) => {
      return cur.count * cur.price + acc;
    }, 0);
    const deliveryFee = 3000;
    return (
      <div className="flex flex-col items-center">
        <span>총가격</span>
        <strong className="text-shoppypink">₩{totalPrice + deliveryFee}</strong>
      </div>
    );
  };

  const productRepository = {
    findAll: jest.fn(),
  };
  const userRepository = {
    updateCount: jest.fn(),
    deleteCarts: jest.fn(),
    findById: jest.fn(),
  };

  afterEach(() => {
    productRepository.findAll.mockReset();
    userRepository.deleteCarts.mockReset();
    userRepository.findById.mockReset();
    userRepository.updateCount.mockReset();
    mockTotalPriceProp.mockReset();
    jest.clearAllMocks();
  });

  it("스냅샷 테스트", async () => {
    productRepository.findAll.mockImplementation(() =>
      Promise.resolve(fakeProducts)
    );
    userRepository.findById.mockImplementation(() =>
      Promise.resolve(fakeAdminUser)
    );

    const { asFragment } = renderCarts();
    await screen.findAllByRole("listitem");
    expect(asFragment()).toMatchSnapshot();
  });

  it("로딩 중이라면 로딩 표시를 해야하며 TotalPrice는 표시되지 말아야한다", () => {
    productRepository.findAll.mockImplementation(() =>
      Promise.resolve(fakeProducts)
    );
    userRepository.findById.mockImplementation(() =>
      Promise.resolve(fakeAdminUser)
    );

    renderCarts();
    expect(screen.getByText("is loading...")).toBeInTheDocument();
    expect(screen.queryByText("총가격")).toBeNull();
  });

  it("에러가 났다면 에러가 났다는 표시를 해야하며 TotalPrice는 표시되지 말아야한다", async () => {
    productRepository.findAll.mockImplementation(() => {
      throw new Error();
    });
    userRepository.findById.mockImplementation(() => {
      throw new Error();
    });
    renderCarts();
    await waitFor(() => {
      expect(screen.getByText("Something is wrong")).toBeInTheDocument();
    });
    await waitFor(() => {
      expect(screen.queryByText("총가격")).toBeNull();
    });
  });

  it("productInCarts가 없다면 CartsCard, TotalPrice는 표시되지 말아야한다", async () => {
    productRepository.findAll.mockImplementation(() => {
      return Promise.resolve(null);
    });
    userRepository.findById.mockImplementation(() =>
      Promise.resolve(fakeAdminUser)
    );
    await waitFor(() => {
      expect(screen.queryAllByRole("listitem")).toHaveLength(0);
    });
    await waitFor(() => {
      expect(screen.queryByText("총가격")).toBeNull();
    });
  });

  it("TotalPrice는 올바른 productsInCarts를 받아야 한다", async () => {
    productRepository.findAll.mockImplementation(() =>
      Promise.resolve(fakeProducts)
    );
    userRepository.findById.mockImplementation(() =>
      Promise.resolve(fakeAdminUser)
    );
    renderCarts();
    await screen.findAllByRole("listitem");
    expect(mockTotalPriceProp).toHaveBeenCalledWith(
      expect.objectContaining({
        1: {
          category: "여성",
          desc: "스타일리쉬한 핑크색이 돋보이는 핫핑크색 후드",
          imageURL: "http://image/",
          name: "핫핑크 후드 HOODY",
          options: "xs,s,m,l,xl",
          price: "40000",
          uid: "abcd",
          count: 1,
          option: "s",
        },
      })
    );
  });

  function renderCarts() {
    return render(
      withAllUserContexts(
        withReactQuery(
          withRouter(
            <Route
              path="/"
              element={<Carts productRepository={productRepository} />}
            />
          )
        ),
        userRepository,
        {
          uid: "1",
          logUserIn: () => {},
        }
      )
    );
  }
});
