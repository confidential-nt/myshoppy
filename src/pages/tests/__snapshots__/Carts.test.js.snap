// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carts 스냅샷 테스트 1`] = `
<DocumentFragment>
  <h1
    class="pt-5 pb-3 font-bold text-center"
  >
    내 장바구니
  </h1>
  <ul
    class="pt-4 pb-4 pl-6 pr-6 border-t border-b w-11/12 mr-auto ml-auto"
  >
    <li
      class="flex items-center mb-2 last:mb-0"
    >
      <div
        aria-label="go-to-product-detail"
        class="w-1/3 mr-4 basis-1/6 cursor-pointer"
        role="img"
      >
        <img
          alt="핫핑크 후드 HOODY"
          class="w-full h-full"
          src="http://image/"
        />
      </div>
      <div
        class="flex justify-between flex-grow basis-2/3"
      >
        <div
          class="flex flex-col justify-center"
        >
          <h3>
            핫핑크 후드 HOODY
          </h3>
          <span
            class="uppercase text-shoppypink font-bold"
          >
            s
          </span>
          <span>
            ₩40000
          </span>
        </div>
        <div
          class="flex"
        >
          <button
            aria-label="deleteCount"
            class="mr-1"
            type="button"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 1024 1024"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M328 544h368c4.4 0 8-3.6 8-8v-48c0-4.4-3.6-8-8-8H328c-4.4 0-8 3.6-8 8v48c0 4.4 3.6 8 8 8z"
              />
              <path
                d="M880 112H144c-17.7 0-32 14.3-32 32v736c0 17.7 14.3 32 32 32h736c17.7 0 32-14.3 32-32V144c0-17.7-14.3-32-32-32zm-40 728H184V184h656v656z"
              />
            </svg>
          </button>
          <span
            class="mr-1 flex items-center"
          >
            1
          </span>
          <button
            aria-label="addCount"
            class="mr-3"
            type="button"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 1024 1024"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M328 544h152v152c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V544h152c4.4 0 8-3.6 8-8v-48c0-4.4-3.6-8-8-8H544V328c0-4.4-3.6-8-8-8h-48c-4.4 0-8 3.6-8 8v152H328c-4.4 0-8 3.6-8 8v48c0 4.4 3.6 8 8 8z"
              />
              <path
                d="M880 112H144c-17.7 0-32 14.3-32 32v736c0 17.7 14.3 32 32 32h736c17.7 0 32-14.3 32-32V144c0-17.7-14.3-32-32-32zm-40 728H184V184h656v656z"
              />
            </svg>
          </button>
          <button
            aria-label="deleteCarts"
            type="button"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 16 16"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M2.5 1a1 1 0 0 0-1 1v1a1 1 0 0 0 1 1H3v9a2 2 0 0 0 2 2h6a2 2 0 0 0 2-2V4h.5a1 1 0 0 0 1-1V2a1 1 0 0 0-1-1H10a1 1 0 0 0-1-1H7a1 1 0 0 0-1 1H2.5zm3 4a.5.5 0 0 1 .5.5v7a.5.5 0 0 1-1 0v-7a.5.5 0 0 1 .5-.5zM8 5a.5.5 0 0 1 .5.5v7a.5.5 0 0 1-1 0v-7A.5.5 0 0 1 8 5zm3 .5v7a.5.5 0 0 1-1 0v-7a.5.5 0 0 1 1 0z"
              />
            </svg>
          </button>
        </div>
      </div>
    </li>
  </ul>
  <div
    class="flex flex-col items-center"
  >
    <span>
      총가격
    </span>
    <strong
      class="text-shoppypink"
    >
      ₩43000
    </strong>
  </div>
</DocumentFragment>
`;
